teerex/src/lib.rs

```diff
@@ -446,6 +500,78 @@ impl<T: Config> Pallet<T> {
 			Err(<Error<T>>::RemoteAttestationTooOld.into())
 		}
 	}
+
+	fn check_block_and_get_latest_header(
+		shard_id: ShardIdentifier,
+		block_header: Header,
+		sender: &T::AccountId,
+		sender_index: u64,
+		latest_block_header: Header,
+	) -> Header {
+		// Confirm that the parent hash is the hash of the previous block.
+		// Block number 1 does not have a previous block, hence skip checking there.
+		if latest_block_header.hash() == block_header.parent_hash || block_header.block_number == 1
+		{
+			Self::confirm_sidechain_block(shard_id, block_header, sender, sender_index);
+			return block_header
+		}
+		latest_block_header
+	}
+
+	fn confirm_sidechain_block(
+		shard_id: ShardIdentifier,
+		block_header: Header,
+		sender: &T::AccountId,
+		sender_index: u64,
+	) {
+		<LatestSidechainBlockHeader<T>>::insert(shard_id, block_header);
+		<WorkerForShard<T>>::insert(shard_id, sender_index);
+		let block_hash = block_header.block_data_hash;
+		log::debug!(
+			"Proposed sidechain block confirmed with shard {:?}, block hash {:?}",
+			shard_id,
+			block_hash
+		);
+		Self::deposit_event(Event::ProposedSidechainBlock(sender.clone(), block_hash));
+	}
+
+	fn import_blocks_from_queue(
+		shard_id: ShardIdentifier,
+		sender: &T::AccountId,
+		sender_index: u64,
+		mut latest_block_header: Header,
+	) {
+		let mut latest_block_number = latest_block_header.block_number;
+		let early_block_proposal_lenience = T::EarlyBlockProposalLenience::get();
+		let mut i: u64 = 0;
+		while <SidechainBlockHeaderQueue<T>>::contains_key(
+			(shard_id, latest_block_number + 1),
```

Comment:
Good catch - there are a couple of unchecked adds in this function: 3x `latest_block_number + 1` and once `i += 1`. The latter should probably be safe, since we only increment until `early_block_proposal_lenience`. But I'm not sure what the best practice is here - maybe all the potential overflow/underflow operations should be checked and therefore also the increment?

Reply:
Substrate seems to have decided to go for the `saturated_add`(https://github.com/paritytech/substrate/blob/master/frame/referenda/src/lib.rs). There you don't have to handle the error, which might be the way to go for a limited number just as `i`